swarmrl/engine/gaurav_sim.py

```diff
@@ -0,0 +1,700 @@
+                mob_A = self._calc_trans_mobility(r_ij_norm, "A")
+                mob_B = self._calc_trans_mobility(r_ij_norm, "B")
+
+                vel_ij_along_r_ij = (
+                    mob_A * (f_mag_on + f_capillary + r_7_force) * r_ij_unit_vec
+                )
+                if edge_edge_dist < 0:
+                    vel_ij_along_r_ij += (
+                        self._calc_trans_mobility(0, "none")
+                        * self.params.raft_repulsion_strength
+                        * edge_edge_dist  # pretty sure this has to be positive
+                        / self.params.raft_radius
+                    )
+
+                vel_ij_along_r_ij_cross_z = mob_B * f_mag_off
+                if edge_edge_dist >= self.params.lubrication_threshold:
+                    r_3_vel = (
+                        -self.params.raft_radius**3 * omegas[i] / r_ij_norm**2
+                    )  # Gaurav: also here which omega
+                    vel_ij_along_r_ij_cross_z += r_3_vel
+                elif self.params.lubrication_threshold > edge_edge_dist > 0:
+                    mob_C = self._calc_trans_mobility(r_ij_norm, "C")
+                    b_field_angle = np.arctan2(b_field[1], b_field[0])
+                    vel_ij_along_r_ij_cross_z += (
+                        mob_C
+                        * mag_moments[i]
+                        * np.linalg.norm(b_field)
+                        * np.sin(b_field_angle - state[i, 2])
+                    )
+                else:
+                    pass
+
+                vels[i, :] += vel_ij_along_r_ij * r_ij_unit_vec
+                vels[i, :] += vel_ij_along_r_ij_cross_z * r_ij_cross_z
+                vels[j, :] += -vel_ij_along_r_ij * r_ij_unit_vec
+                vels[j, :] += -vel_ij_along_r_ij_cross_z * r_ij_cross_z
+
+        return vels
+
+    def get_rhs(self, t, state_flat: np.array):
+        state = state_flat.reshape((-1, 3))
+        b_field = calc_B_field(self.current_action, t)
+        omega = self._get_omega(state, b_field)
+        vel = self._get_vel(state, omega, b_field)
+        state_derivative = np.concatenate([vel, omega[:, None]], axis=1)
+        return state_derivative.reshape(-1)
+
+    def integrate(self, n_slices: int, model: Model):
+
+        if not self.integration_initialised:
+            self.time = 0.0
+            self.slice_idx = 0
+            self._init_h5_output()
+            self.integration_initialised = True
+
+        def rhs(t, state):
+            # get_rhs has self as the first argument so we write this
+            # small wrapper to remove it from the call signature
+            return self.get_rhs(t, state)
+
+        state_flat = self._get_state_from_rafts().flatten()
+
+        n_snapshots_per_slice = int(
+            round(self.params.time_slice / self.params.snapshot_interval)
+        )
+        for _ in range(n_slices):
+            # Check for a model termination condition.
```

Comment:
I added a kill switch here but commented it out as without the main parent classes for the force model it will break the tests.